### ISSUE TITLE
webdav (bearer-)token time out / possible to send custom token

### ISSUE THREAD

**Issue by conrad-stork** [iterative/dvc-webdav#19]
# Bug Report

## webdav (bearer-)token time out

Running `dvc push` ( with a webdav server as remote) results in a timeout error, when authorizing via (bearer-)token.

Here the traceback:
```console
Traceback (most recent call last):
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpcore/_exceptions.py", line 8, in map_exceptions
    yield
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpcore/backends/sync.py", line 26, in read
    return self._sock.recv(max_bytes)
  File "/home/storkc/programms/miniconda3/lib/python3.9/ssl.py", line 1227, in recv
    return self.read(buflen)
  File "/home/storkc/programms/miniconda3/lib/python3.9/ssl.py", line 1102, in read
    return self._sslobj.read(len)
socket.timeout: The read operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpx/_transports/default.py", line 60, in map_httpcore_exceptions
    yield
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpx/_transports/default.py", line 218, in handle_request
    resp = self._pool.handle_request(req)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpcore/_sync/connection_pool.py", line 253, in handle_request
    raise exc
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpcore/_sync/connection_pool.py", line 237, in handle_request
    response = connection.handle_request(request)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpcore/_sync/connection.py", line 90, in handle_request
    return self._connection.handle_request(request)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpcore/_sync/http11.py", line 105, in handle_request
    raise exc
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpcore/_sync/http11.py", line 84, in handle_request
    ) = self._receive_response_headers(**kwargs)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpcore/_sync/http11.py", line 148, in _receive_response_headers
    event = self._receive_event(timeout=timeout)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpcore/_sync/http11.py", line 177, in _receive_event
    data = self._network_stream.read(
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpcore/backends/sync.py", line 26, in read
    return self._sock.recv(max_bytes)
  File "/home/storkc/programms/miniconda3/lib/python3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpcore/_exceptions.py", line 12, in map_exceptions
    raise to_exc(exc)
httpcore.ReadTimeout: The read operation timed out

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/dvc_objects/fs/generic.py", line 310, in transfer
    _try_links(
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/dvc_objects/fs/generic.py", line 249, in _try_links
    return copy(from_fs, from_path, to_fs, to_path, callback=callback)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/dvc_objects/fs/generic.py", line 86, in copy
    return _put(
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/dvc_objects/fs/generic.py", line 146, in _put
    return _put_one(from_paths[0], to_paths[0])
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/dvc_objects/fs/generic.py", line 138, in _put_one
    return put_file(from_path, to_path, callback=callback)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/dvc_objects/fs/callbacks.py", line 69, in func
    return wrapped(path1, path2, **kw)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/dvc_objects/fs/callbacks.py", line 41, in wrapped
    res = fn(*args, **kwargs)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/dvc_objects/fs/base.py", line 534, in put_file
    self.fs.put_file(
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/webdav4/fsspec.py", line 374, in put_file
    return self.upload_fileobj(
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/webdav4/fsspec.py", line 348, in upload_fileobj
    return self.client.upload_fileobj(
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/webdav4/client.py", line 691, in upload_fileobj
    http_resp = self.request(
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/webdav4/client.py", line 383, in request
    http_resp = self._request(method, path, **kwargs)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/webdav4/client.py", line 362, in _request
    http_resp = self.http.request(method, url, **kwargs)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpx/_client.py", line 815, in request
    return self.send(request, auth=auth, follow_redirects=follow_redirects)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpx/_client.py", line 902, in send
    response = self._send_handling_auth(
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpx/_client.py", line 930, in _send_handling_auth
    response = self._send_handling_redirects(
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpx/_client.py", line 967, in _send_handling_redirects
    response = self._send_single_request(request)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpx/_client.py", line 1003, in _send_single_request
    response = transport.handle_request(request)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpx/_transports/default.py", line 218, in handle_request
    resp = self._pool.handle_request(req)
  File "/home/storkc/programms/miniconda3/lib/python3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/home/storkc/programms/miniconda3/lib/python3.9/site-packages/httpx/_transports/default.py", line 77, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.ReadTimeout: The read operation timed out
```

On Server side, we see that requests are coming in, but a PUT statement times out. We are running dvc version 2.43.1. Running the same setup without Auth or with Basic Auth does work.

Note that we are requiring the token for https authentication (via an nginx secured location). The basic auth was directly securing the webdav server.

Further it would be very interesting for us to send also custom headers within a webdav request to our server (we saw that this should be quiet easy to implement [here](https://github.com/iterative/dvc-webdav/blob/main/dvc_webdav/__init__.py#L71)) and we were wondering if there is a reason why this is not possible?

### Reproduce

We have the following setup: Ningx as reverse proxy (which also does authentication) which does a proxy pass to a webdav server (running in an Apache docker container). The Setup is working completely fine when webdav does a simple BasicAuth. Now we have a use case where we want to advance our Auth because more users are using the setup. We want to handle this with custom authorization headers (e.g. bearer authentication) sending with the dvc push requests. This is possible for http already but not for webdav. Other than that we just define our remote and push.

### Expected

It would be nice to have the possibility to send custom headers with webdav requests. Further a timeout using the (bearer-)token should not happen.

### Environment information

```console
DVC version: 2.43.1 (pip)
---------------------------------
Platform: Python 3.9.13 on Linux-5.10.16.3-microsoft-standard-WSL2-x86_64-with-glibc2.31
Subprojects:
        dvc_data = 0.35.1
        dvc_objects = 0.19.0
        dvc_render = 0.0.17
        dvc_task = 0.1.11
        dvclive = 1.3.3
        scmrepo = 0.1.6
Supports:
        http (aiohttp = 3.8.1, aiohttp-retry = 2.7.0),
        https (aiohttp = 3.8.1, aiohttp-retry = 2.7.0),
        webdav (webdav4 = 0.9.7),
        webdavs (webdav4 = 0.9.7)
Cache types: hardlink, symlink
Cache directory: ext4 on /dev/sdb
Caches: local
Remotes: webdavs
Workspace directory: ext4 on /dev/sdb
Repo: dvc, git
```

**Comment by skshetry** [comment]
Hi @conrad-stork, are you able to do `PUT` with curl with the header set?

**Comment by conrad-stork** [comment]
Hi @skshetry, thanks for the quick reply, I had some issues to test it, but now I made it and I can also not doe the PUT method via curl. So it might be a server configuration problem? Any ideas what could go wrong especially for the PUT method? GET method is working...

**Comment by conrad-stork** [comment]
Hi @skshetry again, we now managed to make it work! We had some trouble on the server side. Sorry for bothering you! However could you also comment to the feature request, or should we open a new issue for that? Thanks in advance and Best!

**Comment by skshetry** [comment]
@conrad-stork, the feature request makes sense. If you can create a PR to `dvc-webdav` adding support (and, extending config schema in `dvc`), I'd be happy to review.

**Comment by frederik-sandfort** [comment]
@skshetry I created pull requests for [dvc-webdav](https://github.com/iterative/dvc-webdav/pull/20), [dvc](https://github.com/iterative/dvc/pull/8933) and [dvc.org](https://github.com/iterative/dvc.org/pull/4280). Hope this will be integrated. Thanks a lot for your support!

**Comment by skshetry** [comment]
Closed by #20.
